Create Bidi Level struct type

behnam opened the pull request:
The `Level` struct always represents a valid bidi embedding level, therefore can fail on `new*()` and mutation (`raise()`/`lower()`) methods.

mbrubeck commented on src/level.rs:
```diff
@@ -0,0 +1,258 @@
+}
+
+impl Level {
+    /// New LTR level with smallest number vaule (0)
```
Typo: `s/vaule/value/` (also found in some other comments)

behnam replied:
Thanks, I'll check these out and try one more of proof reading new comments.

mbrubeck commented on src/level.rs:
```diff
@@ -0,0 +1,258 @@
+// Copyright 2014 The html5ever Project Developers. See the
```
This file can have "Copyright 2017."  Also I just noticed that this says "The html5ever Project Developers" which I must have copy-pasted accidentally a long time ago.  Please change it to "The Servo Project Developers" here and elsewhere.  Thanks!

behnam replied:
Sure. Included the second part in the cleanup PR: https://github.com/servo/unicode-bidi/pull/29

mbrubeck commented on src/implicit.rs:
```diff
@@ -193,20 +199,21 @@ pub fn resolve_neutral(sequence: &IsolatingRunSequence, levels: &[u8], classes:
+            (false, AN) | (false, EN) => levels[i].raise(2).expect("Level number error"),
             // http://www.unicode.org/reports/tr9/#I2
-            (true, L) | (true, EN) | (true, AN) => levels[i] += 1,
+            (true, L) | (true, EN) | (true, AN) => levels[i].raise(1).expect("Level number error"),
```
According to the [spec], the level may reach *max_depth* + 1 at this step, so we need different validity checking here than in earlier stages.

[spec]: http://www.unicode.org/reports/tr9/#Resolving_Implicit_Levels

behnam committed:
Create Bidi Level struct type

The `Level` struct always represents a valid bidi embedding level, with
checks for explicit and implicit boundries (max_depth and max_depth+1,
respectively) therefore can fail on `new*()` and mutation
(`raise*()`/`lower()`) methods.

So, we have one level type, which is always a valid number of *implicit
embedding level resolution*, but can be made invalid for *explicit
embedding level resolution*. This is fine because external usage of the
data structure would normally deal with the final resolved levels, which
are the *implicit* ones (and that's why its boundry is made the implicit
boundry here, as well), and the *explicit* bounrdy is supposed to be
checked only during the early stages of the algorithm (the explicit
stages), which can do so explicitly by calling the appropriate methods.

behnam commented:
Okay, I have address the last comment in https://github.com/behnam/rust-unicode-bidi/pull/2/commits/16a40aff2dd3ad0e1bf4ee082afa1bcbdad06f4f . I'm closing the PR and will submit it on the main repo after https://github.com/servo/unicode-bidi/pull/29 gets landed.
